torch/csrc/deploy/deploy.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+#include <assert.h>
+#include <torch/csrc/deploy/interpreter/interpreter_impl.h>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace torch {
+
+struct MovableObject;
+struct InterpreterManager;
+
+struct TORCH_API InterpreterSession {
+  InterpreterSession(InterpreterSessionImpl* impl, InterpreterManager* manager)
+      : impl_(impl), manager_(manager) {}
+
+  PythonObject self; // when retreived from a PythonMovable this will be set.
+  InterpreterSession(InterpreterSession&&) = default;
+  ~InterpreterSession();
+  PythonObject global(const char* module, const char* name) {
+    return impl_->global(module, name);
+  }
+  PythonObject from_ivalue(at::IValue ivalue) {
+    return impl_->from_ivalue(std::move(ivalue));
+  }
+
+  MovableObject create_movable(PythonObject obj);
+
+ private:
+  friend struct MovableObject;
+  friend struct Package;
+  friend struct InterpreterManager;
+  friend struct MovableObjectImpl;
+  std::unique_ptr<InterpreterSessionImpl> impl_;
+  InterpreterManager* manager_; // if created from one
+  int64_t notify_idx_ = -1;
+};
+
+class TORCH_API Interpreter {
+ private:
+  std::string library_name_;
+  void* handle_;
+  std::unique_ptr<InterpreterImpl> pImpl_;
+
+  InterpreterManager* manager_; // optional if managed by one
+
+ public:
+  Interpreter(InterpreterManager* manager);
+  InterpreterSession acquire_session() const {
+    return InterpreterSession(pImpl_->acquire_session(), manager_);
+  }
+  ~Interpreter();
+  Interpreter(Interpreter&& rhs)
+      : library_name_(std::move(rhs.library_name_)),
+        handle_(rhs.handle_),
+        pImpl_(std::move(rhs.pImpl_)),
+        manager_(rhs.manager_) {
+    rhs.handle_ = nullptr;
+  }
+
+  Interpreter(const Interpreter&) = delete;
+  Interpreter& operator=(const Interpreter&) = delete;
+  Interpreter& operator=(Interpreter&&) = delete;
+  friend struct InterpreterManager;
+};
+
+struct Package;
+
+struct TORCH_API LoadBalancer {
+  LoadBalancer(size_t n) : locks_(new uint64_t[8 * n]), n_(n) {
+    // 8*... to avoid false sharing of atomics on the same cache line
+    memset(locks_, 0, 8 * n_ * sizeof(uint64_t));
+  }
+  void setResourceLimit(size_t n) {
+    n_ = n;
+  }
+#ifndef _WIN32
+  int acquire() {
+    thread_local int last = 0;
+    size_t minusers = SIZE_MAX;
+    int min_idx = 0;
+    for (size_t i = 0; i < n_; ++i, ++last) {
+      if (last >= n_) {
+        last = 0;
+      }
+      uint64_t prev =
+          __atomic_fetch_add(&locks_[8 * last], 1ULL, __ATOMIC_SEQ_CST);
+      if (prev == 0) {
+        // fast path, we found an interpreter with no users
+        return last;
+      } else {
+        // slow path, we don't want to use this interpreter because it is being
+        // used by someone else.
+        __atomic_fetch_sub(&locks_[8 * last], 1ULL, __ATOMIC_SEQ_CST);
+      }
```

Comment:
shouldn't this just be a strong CAS where it expects 0 and wants to write 1? (and thus it's not clear to me that this logic is pretty simple...)

Reply:
The counts in `locks_` are just an estimate of how many threads are using each interpreter, they don't have to be strictly accurate in the short time. Ill rename `locks_` because it isn't an accurate name. 

The reason for doing it this way is so that the fast path doesn't have any retry logic -- it just does a single atomic add and can move on to execution. 

While this balancer only really matters for small models, I think this is still important to include because we are selling this as "GIL-free" python election. The very first thing someone might do to evaluate whether this is working will be to run 64 threads, each doing very little work, and checking that they scale. I measured a number of approaches and any that used a single lock or contended atomics scaled for really small programs.

Reply:
Here is what I'm proposing: https://godbolt.org/z/ao6Yod

There isn't retry logic in my proposal either, it's just a lock cmpxchg instead of a lock xadd, and you don't have to do a lock xsub if the interpreter wasn't free.

Reply:
(It's entirely possible I've missed something or that this is actually a bad idea on x86_64 -- I'm not an expert on the hardware primitives, I just saw two atomic operations instead of one.)

Reply:
My thought was that some architectures implement an atomic add faster than a CAS (cuda), but I don't think x86 is one of them. So using a CAS here would be simpler and I can put that in.